### DISCUSSION TITLE
Development and release signature

### DISCUSSION THREAD

**Discussion by bigstefan** [rauc/rauc#1545] [Q&A] [answered]
Hi
Currently we are evaluating RAUC to replace our proprietary updating infrastructure.
In our workflow the auto-builder generate a package/bundle and signs it with the developer key.
After the SW is internally tested and accepted, it will be resigned with the release key.
This release-package will be used for verification and for the customers.

In the chapter [Switching the Keyring – SPKI hashes](https://rauc.readthedocs.io/en/v1.12/advanced.html#switching-the-keyring-spki-hashes) there is a description how to handle this use case.

After read the doc and the script, we remarked that the implementation for the use case applies the following rule. 
If the board runs a developer-signed SW, it can receive a developer-signed bundle and flash it.
If the board runs a release-signed SW, it can receive a release-signed bundle and flash it.
So it 'inherit' the signature to check (the bundle contains both, and select the cert in the post-process-hook).

In our use case we like to able to receive a developer-signed and a release-signed bundle on a developer-signed SW.
Today when we receive a package we, try to check the release-signature, if this fails and we are running on a 'developer-enabled' board we also check the developer-signature.
We have flag in the secure place to define if the board is 'developer-enabled' or not.

Is there any possibility to support such a behavior?
We are not seeing any, this is why we ask here.

Thanks for your help
Best Regards
Stefan

**Comment by jluebbe** [comment, marked as answer]
The keyring file can contain multiple public keys, all of which are then accepted during authentication.

> Today when we receive a package we, try to check the release-signature, if this fails and we
> are running on a 'developer-enabled' board we also check the developer-signature.
> We have flag in the secure place to define if the board is 'developer-enabled' or not

That sounds like you **don't** want to switch a given board between 'release' and 'developer-enabled' during installation.
The approach of using the SPKI hashes is only relevant if you want to switch depending on how the bundle was signed.

Instead, the keyring selection should depend on this 'developer-enabled' flag. You'd have two keyrings:
- `devel-keyring.pem` (containing both certificates)
- `release-keyring.pem` (containing only the release keyring)

To put the correct keyring in place, I see two simple ways:
- in a `slot-post-install` hook, check your custom flag and switch based on that (instead based on SPKI hashes)
- on boot, based on the flag, create a symlink or bind mount to make the correct keyring accessible to RAUC

**Reply by bigstefan** [reply]
Hi Jan

Thanks a lot for this answer, it's obvious.
I will implement as you proposed.

Regards Stefan
